### ISSUE TITLE
Error trying to subscribe to 100 currency markets -  is there a limit?

### ISSUE THREAD

**Issue by superqd** [gliwka/bittrex-market#17]
I am trying to subscribe to 100  currency markets, to get the orderBookUpdated events for each.  Here is my code

```
var MarketManager = require("bittrex-market");

var marketDictionary = {};
marketDictionary["BTC-LTC"] = true;
marketDictionary["BTC-DOGE"] = true;
marketDictionary["BTC-VTC"] = true;
marketDictionary["BTC-PPC"] = true;
marketDictionary["BTC-FTC"] = true;
// ... more dictionary entries about 100 total

var marketManager = new MarketManager(false);
var init() = function(){
    for(var market in marketDictionary){        
        marketManager.market(market, function(error, m){        
            m.on("orderbookUpdated", function(){
                console.log(market +": asks: "+ m.asks.length +", bids: "+ m.bids.length);
            });
        });
    }
};

init();
```

When I run this, I get the following error:

```
...\node_modules\bittrex-market\lib\manager.js:135
        if(this._subscribing || !client.wasConnected) {
                                       ^
TypeError: Cannot read property 'wasConnected' of null
```

I'd seen some info about the number of event listeners, but I'm not sure that is the issue.  Any ideas what is the problem?  Is there a limit to the number of currency markets you can subscribe to?  I'd seen another open issue with answers that seemed to imply there is no limit, but I'm having trouble with 100.

Actually, I just commented out all but two of the entries and I get the same error.  It only worked if the marketDictionary had a single entry.  I tried this instead:

```
var managers = {};
var init = function(){
    for(var market in marketDictionary){
        var manager = managers[market] = new MarketManager(false);
        
        manager.market(market, function(error, m){        
            m.on("orderbookUpdated", function(){
                console.log(market +": asks: "+ m.asks.length +", bids: "+ m.bids.length);
            });
        });
    }
};
```

And this too failed spectacularly.  It said the cloudflare protection could not be circumvented, and on another time it returned some indecipherable error.  The point is, even instantiating a new manager for each market seems not to work.  Here was the error:

```
\node_modules\bittrex-market\lib\manager.js:102
                throw "Failed to circumvent cloudflare protection: " + error
                ^
Failed to circumvent cloudflare protection: [object Object]
```

How do you subscribe to multiple currencies?  Is that possible?

**Comment by gliwka** [comment]
Thanks for the bug report - I'll take a look!

**Comment by superqd** [comment]
I had a chance to look at this tonight.  The problem is in **manager.js**.  The second call to **_addClient()** is occurring before the call to **cloudscraper.get()** can complete.  That is, this code:

```
_addClient(callback) {
        if(this._subscribedCount != 1 && this._subscribedCount % (maxCurrenciesPerClient + 1) != 0) {
            return callback(this._currentClient)
        }
```

is getting executed the second time, before this code can finish:

```
cloudscraper.get('https://bittrex.com/', (error, response, body) => {
            if(error) {
                throw "Failed to circumvent cloudflare protection: " + error
            }

            this._cloudflareUserAgent = response.request.headers['User-Agent']
            this._cloudflareCookies = response.request.headers['cookie']
            this._cloudflareBypassCallbacksWaiting.forEach(savedCallback => {
                savedCallback()
            })
        })
```
And since the **saveCallback()** is actually what gets called to *create* the **_currentClient**, the second call to **_addClient()** returns a null client value.

Also, the cloudscraper.get() code needs to have this added at the end:

`this._cloudflareBypassRequestInFlight = false`

Otherwise it will only ever execute once.

**Comment by gliwka** [comment]
Thanks for looking into this, I'm currently moving between continents (Europe to Asia), so do not have a lot of time on my hands :-( I'll take a look at this in the coming week when I arrive at my destination!

**Comment by superqd** [comment]
Best of luck!

**Comment by gliwka** [comment]
Thanks!

**Comment by jevdiego** [comment]
Hi @gliwka Any chance you've had some time to look at this?

**Comment by nouseforacode** [comment]
```
const marketManager = new MarketManager(false)
marketManager._cloudflareBypassReady = true
```

this works for me

**Comment by jevdiego** [comment]
Fantastic... that works for me too!

**Comment by superqd** [comment]
Yeah, that got things working for me as well!  Here is my updated code:

```
var marketManager = new MarketManager(false);
marketManager._cloudflareBypassReady = true;

var init2 = function(){
    for(var market in marketDictionary){        
        (function(market){
            marketManager.market(market, function(error, m){    
                if(!m) return;
                m.on("orderbookUpdated", function(){
                    console.log(market +": asks: "+ m.asks.length +", bids: "+ m.bids.length);
                });
            });
        })(market);  
    }
};
```


Where marketDictionary is of this form:

```
var marketDictionary = {};
marketDictionary["BTC-LTC"] = true;
marketDictionary["BTC-DOGE"] = true;
marketDictionary["BTC-VTC"] = true;
marketDictionary["BTC-PPC"] = true;
marketDictionary["BTC-FTC"] = true;
marketDictionary["BTC-RDD"] = true;
marketDictionary["BTC-NXT"] = true;
marketDictionary["BTC-DASH"] = true;
marketDictionary["BTC-POT"] = true;
marketDictionary["BTC-BLK"] = true;
marketDictionary["BTC-EMC2"] = true;
...
```

And has about 114 or so entries.

Thanks for the suggestion, @nouseforacode !

**Comment by gliwka** [comment]
Sorry for the inconvinience caused by this bug, it's now fixed in v2.2.0.

**Comment by superqd** [comment]
No worries.  Thanks for following up!
